File: backend/app/models/goal_discovery.py
"""
Goal discovery (motivators + notification profile) models.

These models power a back-and-forth conversation that helps the system
understand what matters to the user and how to nudge them.
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional, List, Literal

from pydantic import BaseModel, Field


NotificationStyle = Literal["gentle", "direct", "playful", "mixed"]


class NotificationProfile(BaseModel):
    """Structured, persisted profile used to personalize nudges/notifications."""

    user_id: str

    # Identity + goal framing
    identity: Optional[str] = Field(
        None, description="Identity label the user resonates with (e.g., 'writer')"
    )
    primary_goal: Optional[str] = Field(
        None, description="The single most important goal right now"
    )
    why: Optional[str] = Field(None, description="Why the goal matters to them")

    # Motivation + meaning
    motivators: List[str] = Field(default_factory=list)
    stakes: Optional[str] = Field(
        None, description="What it would cost them emotionally/practically to give up"
    )
    importance_1_to_5: Optional[int] = Field(
        None, ge=1, le=5, description="How much this goal matters to them (1-5)"
    )

    # Notification preferences
    style: NotificationStyle = "gentle"
    preferred_name_for_user: Optional[str] = None
    preferred_name_for_assistant: Optional[str] = None

    # App-specific intent/rules
    helpful_apps: List[str] = Field(default_factory=list)
    risky_apps: List[str] = Field(default_factory=list)
    app_intent_notes: Optional[str] = Field(
        None,
        description="Freeform notes like 'YouTube helps when learning writing, hurts when doomscrolling'",
    )

    updated_at: datetime = Field(default_factory=datetime.utcnow)


class GoalDiscoveryStartRequest(BaseModel):
    """Start or reset a goal-discovery session."""

    reset: bool = False


class GoalDiscoveryMessageRequest(BaseModel):
    """Send a message within a goal-discovery session."""

    session_id: str = Field(..., min_length=1)
    message: str = Field(..., min_length=1, max_length=2000)


class GoalDiscoveryResponse(BaseModel):
    """Response from goal-discovery assistant."""

    session_id: str
    message: str
    done: bool = False
    profile: Optional[NotificationProfile] = None

File: lib/models/goal_discovery.dart
/// Models for the goal discovery / notification profile flow.

class NotificationProfile {
  final String? identity;
  final String? primaryGoal;
  final String? why;
  final List<String> motivators;
  final String? stakes;
  final int? importance1To5;
  final String? style;
  final String? preferredNameForUser;
  final String? preferredNameForAssistant;
  final List<String> helpfulApps;
  final List<String> riskyApps;
  final String? appIntentNotes;

  const NotificationProfile({
    this.identity,
    this.primaryGoal,
    this.why,
    this.motivators = const [],
    this.stakes,
    this.importance1To5,
    this.style,
    this.preferredNameForUser,
    this.preferredNameForAssistant,
    this.helpfulApps = const [],
    this.riskyApps = const [],
    this.appIntentNotes,
  });

  factory NotificationProfile.fromJson(Map<String, dynamic> json) {
    return NotificationProfile(
      identity: json['identity'],
      primaryGoal: json['primary_goal'],
      why: json['why'],
      motivators: (json['motivators'] as List<dynamic>? ?? [])
          .map((e) => e.toString())
          .toList(),
      stakes: json['stakes'],
      importance1To5: json['importance_1_to_5'],
      style: json['style'],
      preferredNameForUser: json['preferred_name_for_user'],
      preferredNameForAssistant: json['preferred_name_for_assistant'],
      helpfulApps: (json['helpful_apps'] as List<dynamic>? ?? [])
          .map((e) => e.toString())
          .toList(),
      riskyApps: (json['risky_apps'] as List<dynamic>? ?? [])
          .map((e) => e.toString())
          .toList(),
      appIntentNotes: json['app_intent_notes'],
    );
  }
}

class GoalDiscoveryResponse {
  final String sessionId;
  final String message;
  final bool done;
  final NotificationProfile? profile;

  const GoalDiscoveryResponse({
    required this.sessionId,
    required this.message,
    required this.done,
    this.profile,
  });

  factory GoalDiscoveryResponse.fromJson(Map<String, dynamic> json) {
    return GoalDiscoveryResponse(
      sessionId: json['session_id'] ?? '',
      message: json['message'] ?? '',
      done: json['done'] ?? false,
      profile: json['profile'] != null
          ? NotificationProfile.fromJson(
              json['profile'] as Map<String, dynamic>,
            )
          : null,
    );
  }
}



